sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinations.java

```diff
@@ -0,0 +1,159 @@
+        throws IllegalStateException {
+      @SuppressWarnings("unchecked")
+      SideInputT materialized = (SideInputT)  processContext.sideInput(sideInput);
+      return materialized;
+    }
+  }
+
+  /**
+   * Specifies that this object needs access to one or more side inputs. This side input must be
+   * globally windowed, as it will be accessed from the global window.
+   */
+  public List<PCollectionView<?>> getSideInputs() {
+    return Lists.newArrayList();
+  }
+
+  /**
+   * Returns an object that represents at a high level which table is being written to. May not
+   * return null.
+   */
+  public abstract DestinationT getDestination(ValueInSingleWindow<T> element);
+
+  /**
+   * Returns the coder for {@link DestinationT}. If this is not overridden, then
+   * {@link BigQueryIO} will look in the coder registry for a suitable coder. This must be a
+   * deterministic coder, as {@link DestinationT} will be used as a key type in a
+   * {@link org.apache.beam.sdk.transforms.GroupByKey}.
+   */
+  @Nullable
+  public Coder<DestinationT> getDestinationCoder() {
+    return null;
+  }
+
+  /**
+   * Returns a {@link TableDestination} object for the destination. May not return null.
+   */
+  public abstract TableDestination getTable(DestinationT destination,
+                                            SideInputAccessor sideInputAccessor);
+
+  /**
+   * Returns the table schema for the destination. May not return null.
+   */
+  public abstract TableSchema getSchema(DestinationT destination,
+                                        SideInputAccessor sideInputAccessor);
+
+
+  // Gets the destination coder. If the user does not provide one, try to find one in the coder
+  // registry. If no coder can be found, throws CannotProvideCoderException.
+  Coder<DestinationT> getDestinationCoderWithDefault(CoderRegistry registry)
+      throws CannotProvideCoderException {
+    Coder<DestinationT> destinationCoder = getDestinationCoder();
+    // If dynamicDestinations doesn't provide a coder, try to find it in the coder registry.
+    // We must first use reflection to figure out what the type parameter is.
+    Type superclass = getClass().getGenericSuperclass();
+    while (destinationCoder == null) {
+      if (superclass instanceof ParameterizedType) {
+        ParameterizedType parameterized = (ParameterizedType) superclass;
+        if (parameterized.getRawType() == DynamicDestinations.class) {
+          // DestinationT is the second parameter.
+          Type parameter = parameterized.getActualTypeArguments()[1];
+          @SuppressWarnings("unchecked")
+          Class<DestinationT> parameterClass = (Class<DestinationT>) parameter;
+          destinationCoder = registry.getDefaultCoder(parameterClass);
+          if (destinationCoder == null) {
+            throw new CannotProvideCoderException("Could not find a coder for destination type "
+                + parameterClass);
+          }
+          break;
```

Comment:
s/break/return destinationCoder;
and replace line 139 with while(superclass != null) and line 157 with a throw.